### DISCUSSION TITLE
WaitForNetIP and docker-backed vcsim VM

### DISCUSSION THREAD

**Discussion by BrianMMcClain** [vmware/govmomi#3020] [Community Help] [answered]
Howdy!

This question may just be me overlooking something, so I apologize if I am :)

The tl;dr is that I'm experimenting with [backing a VM with a docker container](https://github.com/vmware/govmomi/wiki/vcsim-features#containers-as-vms) to integrate with another existing product and ran into a small speedbump with the call to `WaitForNetIP`, where it hangs until timeout. Alternatively, `WaitForIP` returns the correct IP immediately. In my case, the VM for both calls has one NIC

`WaitForNetIP`, if my understanding is correct, does the following...

1. pulls the `config.hardware.device` of the VM
2. stores a list of MAC addresses for each `BaseVirtualEthernetCard` 
3. pulls the `guest.net` property of the VM
4. Iterates over each NIC, comparing them to the list of MAC addresses stored in step 2
5. Stores the IP for each found MAC

(I think I'm over simplifying it, and I could be misunderstanding here)

The first thing to address was that the MAC reported by `config.hardware.device` and `guest.net` differ, which makes sense reading the wiki, I saw:

> The VM guest field is populated with the container's IP address and MAC address

The MAC reported from the VM is VMware MAC, while the one from the guest.net property is from Docker. 

However, if I create a VM where the MACs match, it still continues to hang. I believe the spot I'm struggling with right now is this block in [virtual_machine.go](https://github.com/vmware/govmomi/blob/68d1994e444f822644315919dab87553d63b46f8/object/virtual_machine.go#L378)

```go
if mac == "" || nic.IpConfig == nil {
    continue
}
```

where `mac` is populated correctly, but `nic.IpConfig` is nil, so it skips the following block which parses and stores the IP address from the NIC.

I figured it wouldn't hurt to drop a message here and see if this scenario is a supported one, or if this field isn't populated from vcsim. If the latter, would this be something that would be a good PR? I'm still reading through the code so haven't seen why this is the case, but happy to follow through. Thanks!

**Comment by dougm** [comment, marked as answer]
Hey @BrianMMcClain !  Thanks for the details and sorry for the delay. The patch below should fix, I'll update the tests and open a PR soon (need to resolve another issue first).

```diff
diff --git a/simulator/container.go b/simulator/container.go
index 08f8f7ed..650c1997 100644
--- a/simulator/container.go
+++ b/simulator/container.go
@@ -114,6 +114,18 @@ func (c *container) inspect(vm *VirtualMachine) error {
                        net := &vm.Guest.Net[0]
                        net.IpAddress = []string{s.IPAddress}
                        net.MacAddress = s.MacAddress
+                       net.IpConfig = &types.NetIpConfigInfo{
+                               IpAddress: []types.NetIpConfigInfoIpAddress{
+                                       {IpAddress: s.IPAddress},
+                               },
+                       }
+               }
+
+               for _, d := range vm.Config.Hardware.Device {
+                       if eth, ok := d.(types.BaseVirtualEthernetCard); ok {
+                               eth.GetVirtualEthernetCard().MacAddress = s.MacAddress
+                               break
+                       }
                }
        }
```

**Reply by BrianMMcClain** [reply]
Oh this is awesome, thank you! I'll give this a spin and let you know how it goes :)

**Reply by BrianMMcClain** [reply]
Great news, this was exactly what I needed, thank you! :)

**Reply by dougm** [reply]
Nice! PR opened: https://github.com/vmware/govmomi/pull/3034
